docs: state management (#8547)
sveltejs/kit: documentation/docs/20-core-concepts/50-state-management.md

---
title: State Management
---

Managing state is one of the hardest parts in application development. This section covers various use cases with regards to state management and what to watch out for.

## Avoid global state in SSR

> If you are creating an [SPA](glossary#csr) with SvelteKit, you can create global state freely, as you can be sure that it's only initialized inside the user's browser. If you use [SSR](glossary#ssr) however, you have to watch out for a couple of things when managing state. In many server environments, a single instance of your app will serve multiple users (this is not specific to SvelteKit - it's one of the gotchas of working with such environments). For that reason, per-request or per-user state must not be stored in global variables.

Reviewer: I read SPA as the word "spa", so "an spa" sounds weird to me - though it'd be correct for folks reading it as s-p-a. Maybe we just write out "single page application" to avoid the issue
```suggestion
If you are creating a [single page application (SPA)](glossary#csr) with SvelteKit, you can create global state freely, as you can be sure that it's only initialized inside the user's browser. If you use [SSR](glossary#ssr) however, you have to watch out for a couple of things when managing state. In many server environments, a single instance of your app will serve multiple users (this is not specific to SvelteKit - it's one of the gotchas of working with such environments). For that reason, per-request or per-user state must not be stored in global variables.
```

Author: I linked there so people can look up good to enter spa mode

Reviewer: doesn't `glossary#spa` better describe how to enter SPA mode than the CSR entry does? Neither says directly, but the SPA entry directs users to `adapter-static#spa-mode`

Consider the following example where the user is set from inside a `load` function:

```js 
/// file: +page.js
// @filename: ambient.d.ts
declare module '$lib/user' {
	export const user: { set: (value: any) => void };
}

// @filename: index.js
// ---cut---
// DON'T DO THIS!
import { user } from '$lib/user';

/** @type {import('./$types').PageLoad} */
export async function load({ fetch }) {
	const response = await fetch('/api/user');
	user.set(await response.json());
}
```

If you are using SSR, the `load` function will run on the server initially, which means that the whole server instance which serves _all_ requests from _all_ users has its `user` state set to the one just requested from the API. To scope this to a single user, you have a couple of options:

- if you need to access the user state only inside server `load` functions, use [`locals`](hooks#server-hooks-handle)

> - if you need to persist the user state across reloads, but only need to access it inside `load` functions, use [`cookies` in server `load` functions](load#cookies-and-headers)

Reviewer: cookies aren't available only in load functions, but also in api endpoints, etc. also, you'd often not store the state in a cookie, but a key to lookup the state

> - if you need to access and update the state inside components, use Svelte's [context feature](https://svelte.dev/docs#run-time-svelte-setcontext). That way, the state is scoped to components, which means they are not shared across different requests on the server. The drawback is that you can only set and subscribe to the store at component initialization. SvelteKit's stores from `$app/stores` for example are setup like this (which is why you may have encountered a related error message)

Reviewer: users may not know what "component initialization" means. maybe we can clarify where it can be used (e.g. in a `<script>` tag, but not reactive statement - I don't know if that's exactly right but just to give an idea)

---

> title: State Management

Reviewer 2: ```suggestion
title: State management
```

```svelte
/// +layout.svelte
<script>
	import { setContext } from 'svelte';

	/** @type {import('./$types').LayoutData} */
	export let data;

	// Create a store...
	const user = writable(data.user);
	// ...add it to the context for child components to access
	setContext('user', user);
	// Optionally update the data everytime the load function is rerun
	$: $user = data.user;
</script>
```

```svelte
/// +src/user/+page.svelte
<script>
	import { getContext } from 'svelte';

	// Retrieve user store from context
	const user = getContext('user');
</script>

<p>Welcome {$user.name}</p>
```

If you have global data whose initial state is not dependent on a request (in other words, it's always the same), then you can keep storing that data globally, as long as you make sure you don't update it during the initial rendering on the server (during load or component render).

> ## Managing forms
> 
> When coming from a pure Svelte or JavaScript background, you might be used to handling all form interactions through JavaScript. This works well when JavaScript is available but results in unresponsive UI when it isn't (which may be [more often than you think](https://kryogenix.org/code/browser/everyonehasjs.html)). If this is a concern to you, leverage SvelteKit's [form actions](form-actions) instead.

Reviewer 2: Does this belong here? It's well covered by the form section, and I don't think of mutations as part of state management

```suggestion
```

Author: Mutating state is an integral part of managing state to me. If all state was just readonly, managing it would be a lot easier 😄
